### PR TITLE
Refactor - break LedgerState into smaller modules

### PR DESCRIPTION
**Pull request by Soupstraw**
This PR breaks the `Cardano.Ledger.Shelley.LedgerState` into smaller submodules.

Removed `WitHashes` and replaced all references to it with `Set (KeyHash 'Witness (Crypto era))`. 
~~**^ This might break some things upstream**~~ 
(Added deprecated functions and types to make everything backwards-compatible)

Before:
Before
After:
modules png-1

### CODE REVIEW

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState.hs [hunk 1]
```diff
@@ -28,272 +20,78 @@
 -- as state transformations on a ledger state ('LedgerState'),
 -- as specified in /A Simplified Formal Specification of a UTxO Ledger/.
 module Cardano.Ledger.Shelley.LedgerState
-  ( AccountState (..),
```
Comment (lehins) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState.hs]: In order to reduce breakage downstream we still need to make sure we export everything from the LedgerState module that we exported before.

File: eras/shelley/impl/cardano-ledger-shelley.cabal [hunk 2]
```diff
@@ -50,6 +50,10 @@ library
     Cardano.Ledger.Shelley.Genesis
     Cardano.Ledger.Shelley.HardForks
     Cardano.Ledger.Shelley.LedgerState
+    Cardano.Ledger.Shelley.LedgerState.Types
+    Cardano.Ledger.Shelley.LedgerState.DPState
+    Cardano.Ledger.Shelley.LedgerState.PulsingReward
+    Cardano.Ledger.Shelley.LedgerState.IncrementalStake
```
Comment (lehins) [inline eras/shelley/impl/cardano-ledger-shelley.cabal]: Maybe it is better to make those modules internal and let everything LedgerState related be exported from `Cardano.Ledger.Shelley.LedgerState`, just as it was before.

Less modules for public to deal with is better.

**Review by lehins — CHANGES_REQUESTED** [review]
I did a quick skim through the PR and it overall looks pretty good. I made couple of comments on the re-exporting.

<strike> Also I do like having some visualizations and stuff, however they are bound to get outdated over time. So I don't think it is a good idea to add this image into the repo: `libs/cardano-ledger-core/modules.png-1.png` The only way we could add such graphs to the repo if we have a CI action that validates they did not go out of sync together with a clear readme section on how to regenerate them when they do.</strike> I saw that you removed it :smile:, but it was only after I submitted the review

One more important point. I am sure you know this, but just to be safe I'd like to mention it anyways, make sure you don't add any semantic changes in this PR, since in refactorings like that it is hard to notice any changes to the logic implemented in functions.

**Comment by JaredCorduan** [comment]
Note that this PR closes  #1675

File: eras/shelley/impl/cardano-ledger-shelley.cabal (hunk 2 above)
Comment (Soupstraw) [reply eras/shelley/impl/cardano-ledger-shelley.cabal]: Done, moved the submodules to the `other-modules` section.

**Review by JaredCorduan — APPROVED** [review]
I really like all these changes!

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState.hs (hunk 1 above)
Comment (Soupstraw) [reply eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState.hs]: Done, and I added deprecated functions and a type synonym for WitHashes.

**Review by lehins — APPROVED** [review]
Looks awesome. Love this cleanup

**Pull request merged**